File: majority-element.js
//
// https://leetcode.com/problems/majority-element/
//

// One pass hash table
var majorityElement = function(nums) {
    var hashTable = {};
    var majorityMin = nums.length / 2;

    for (var i = 0; i < nums.length; i++){
        current = nums[i];
        count = hashTable[current] || 0;
        count += 1;

        if (count > majorityMin) {
            return current
        }

        hashTable[current] = count;

    }
};


// Boyer-Moore Voting Algorithm
var majorityElementBoyerMoore = function(nums) {
    var count = 0;
    var candidate = null;

    for (var i = 0; i < nums.length; i++){
        let current = nums[i];

        if (count === 0){
            candidate = current;
        }

        let toAdd;
        if (candidate === current){
            toAdd = 1;
        } else {
            toAdd = -1;
        }

        count += toAdd;
    }
    return candidate
};

File: majority-element.py
"""
https://leetcode.com/problems/majority-element/
"""


# Two pass hash table
def majorityElement(nums):
    hash_table = {}
    majority_min = len(nums) / 2

    for num in nums:
        count = hash_table.get(num, 0)
        count += 1
        hash_table[num] = count

    for key, value in hash_table.items():
        if value > majority_min:
            return key


# One pass hash table
def majorityElementOnePass(nums):
    hash_table = {}
    majority_min = len(nums) / 2

    for num in nums:
        count = hash_table.get(num, 0)
        count += 1

        if count > majority_min:
            return num

        hash_table[num] = count


# Boyer-Moore voting algorithm - this is so cool!
def majorityElementBoyerMoore(self, nums):
    count = 0
    candidate = None

    for num in nums:
        if count == 0:
            candidate = num

        count += 1 if num == candidate else -1

    return candidate
